acceptance: test rapid restarts and early queries

tbg opened the pull request:
See #19559, #19579, #19571 for issues touched by this test (all of which
it reproduced before the preceding commits fixed them).

The changes here (outside of the test) ensure that a single-node instance starts
as quickly as possible, without waiting for ports to be ready or for replication
to have occurred. This was done to remove any additional delay in
`TestRapidRestarts` that could otherwise obscur bugs (for example, consider that
going through `./cockroach quit` was not able to tickle any of the bugs).

I was also tempted to "bump" the version upgrade tests to use 1.1 instead of 1.0
since that would have removed a bunch of cruft (since 1.0 doesn't write its
listener files), but I decided against that. As a result, some of the new
accessors look a little unclean, but they should straighten themselves out once
we only rely on listener files.

tbg committed:
server: avoid Fatal when draining early

Fixes #19579.

tbg committed:
server: don't drain partially started server

Fixes #19571.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19591)
<!-- Reviewable:end -->

a-robinson commented:
I didn't review the last commit yet because you said not to.

> Fixes #19493.

That's a PR -- which issue did you mean?

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: 2 of 9 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cli/start.go, line 694 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxU4lhph9lMD9g9hv4R:-KxU4lhph9lMD9g9hv4S:bvqxpls) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e30508862074d42057f13e10938448a1ba2454/pkg/cli/start.go#L694)):*
> ```Go
> 			serverStatusMu.Lock()
> 			serverStatusMu.draining = true
> 			needToDrain := serverStatusMu.started
> ```

I might change this from `needToDrain` to `drainingIsSafe` or something along those lines, since this is changing the logic from aggressively draining whenever the server might have started to only draining if we're positive the server has started (and potentially not draining even if it has started).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-:-KxU3na6BKRjbuxmRbqn:boatl4s)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
#19579, updated the message.

---

Review status: 2 of 9 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cli/start.go, line 694 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxU4lhph9lMD9g9hv4R:-Kxdi5IrelzKpyjw2u5F:bz7gxie) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e30508862074d42057f13e10938448a1ba2454/pkg/cli/start.go#L694)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I might change this from `needToDrain` to `drainingIsSafe` or something along those lines, since this is changing the logic from aggressively draining whenever the server might have started to only draining if we're positive the server has started (and potentially not draining even if it has started).
</blockquote></details>

Makes sense, renamed to `drainingIsSafe`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-:-KxdiLcfwj_imqYZMhAR:b4rqq2o)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Pushed an update, review to your heart's content! The test has already found an additional error, but it'll take some looking into.

tbg committed:
server: don't close the stopper unless draining is safe

This fixes a panic that would occur if the RocksDB instance was closed
with moving parts still trying to use it.

tbg committed:
acceptance: add TestRapidRestarts

See #19559, #19579, #19571 for issues touched by this test (all of which
it reproduced before the preceding commits fixed them).

The changes here (outside of the test) ensure that a single-node instance starts
as quickly as possible, without waiting for ports to be ready or for replication
to have occurred. This was done to remove any additional delay in
`TestRapidRestarts` that could otherwise obscur bugs (for example, consider that
going through `./cockroach quit` was not able to tickle any of the bugs). #
Please enter the commit message for your changes. Lines starting # with '#' will
be ignored, and an empty message aborts the commit.

I was also tempted to "bump" the version upgrade tests to use 1.1 instead of 1.0
since that would have removed a bunch of cruft (since 1.0 doesn't write its
listener files), but I decided against that. As a result, some of the new
accessors look a little unclean, but they should straighten themselves out once
we only rely on listener files.

nvanbenschoten commented:
Reviewed 1 of 1 files at r2, 8 of 8 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 6 of 7 files at r8.
Review status: 8 of 9 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/rapid_restart_test.go, line 33 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-Kxf_HJu_rayJqwiPU8w:-Kxf_HJve3KXdyKFEyeN:b-exf04b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L33)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/log"
> 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
> 	"github.com/pkg/errors"
> ```

move up

---

*[pkg/acceptance/rapid_restart_test.go, line 71 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfaiF83icnIhEl259O:-KxfaiF83icnIhEl259P:b-jzkq5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L71)):*
> ```Go
> 		// -1 corresponds to "interrupted", i.e. receiving SIGINT before setting up
> 		// our own handlers, and 1 corresponds to receiving SIGINT after.
> 		if status := s.ExitStatus(); status != -1 && status != 1 {
> ```

nit: I think this could be cleaned up using the following structure.
```
status := s.ExitStatus()
switch status {
case -1:
    // -1 corresponds to ...
case 1:
    // 1 corresponds to ...
default:
    return ...
}
```

---

*[pkg/acceptance/rapid_restart_test.go, line 84 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfcbQV9Z_wNjRyivxv:-KxfcbQV9Z_wNjRyivxw:b-viteno) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L84)):*
> ```Go
> 	// LocalCluster that skip all the waiting so that we get to kill the process
> 	// early in its boot sequence.
> 	cfg.Nodes = cfg.Nodes[:1]
> ```

Do you think there's value in having one of these tests for a multi-node cluster?

---

*[pkg/acceptance/rapid_restart_test.go, line 103 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfbUhHk5xK3CepCWvg:-KxfbUhHk5xK3CepCWvh:b8afh8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L103)):*
> ```Go
> 		// would get exit code 255, typically.
> 		//
> 		// FIXINPR(tschottdorf): real signal handling (platform specific, the below is darwin).
> ```

Isn't this TODO already covered/discussed in `unexpectedExitCode`?

---

*[pkg/acceptance/rapid_restart_test.go, line 122 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfcNDacObL0SixWEno:-KxfcNDacObL0SixWEnp:b63hg3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L122)):*
> ```Go
> 					base := c.URL(ctx, 0)
> 					if base != "" {
> 						// Torture the prometheus endpoint to prevent regression of #19559.
> ```

We don't want to do this for fresh nodes?

---

*[pkg/acceptance/version_upgrade_test.go, line 69 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfcjsOzIj4aRpNuqjD:-KxfcjsOzIj4aRpNuqjE:becfln0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/version_upgrade_test.go#L69)):*
> ```Go
> 	// Verify that the nodes are *really* at the versions configured. This
> 	// tests the CI harness.
> 	log.Info(ctx, "verify that configured versions are actual versions")
> ```

super nit: verify -> verifying

---

*[pkg/acceptance/localcluster/cluster.go, line 706 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfdQEG60cuzOMsawoB:-KxfdQEG60cuzOMsawoC:b-zdkbhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/localcluster/cluster.go#L706)):*
> ```Go
> }
> 
> func addrFromFile(f string) string {
> ```

nit: this should probably have a more general name, like `readFileIfExists`

---

*[pkg/cli/start.go, line 656 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfZ4bNdoIx8tjszSny:-KxfZ4bNdoIx8tjszSnz:b-y1xg4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/272bc0b4fe655531e8572848e695313d23028a37/pkg/cli/start.go#L656)):*
> ```Go
> 	var returnErr error
> 
> 	noDrain := make(chan struct{}) // closed if interrupted very early
> ```

`noDrain` doesn't convey very much meaning when only reading the bottom of this function. Consider `stopWithoutDrain`.

---

*[pkg/cli/start.go, line 698 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfZ_ewEomxniI0vrGF:-KxfZ_ewEomxniI0vrGG:b-rn5nag) ([raw file](https://github.com/cockroachdb/cockroach/blob/272bc0b4fe655531e8572848e695313d23028a37/pkg/cli/start.go#L698)):*
> ```Go
> 			drainingIsSafe := serverStatusMu.started
> 			serverStatusMu.Unlock()
> 			if !drainingIsSafe {
> ```

Add a comment that checking `drainingIsSafe` here can race with it being set but that it's ok to not drain even if the `Server` has started (it's just not desirable).

---

*[pkg/cli/start.go, line 754 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfZJFG3SEp6IhGzJQ3:-KxfZJFG3SEp6IhGzJQ4:b-kj2oar) ([raw file](https://github.com/cockroachdb/cockroach/blob/272bc0b4fe655531e8572848e695313d23028a37/pkg/cli/start.go#L754)):*
> server drained ...

did it? Your call if it's worth printing a different message here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 7 of 7 files at r4, 1 of 1 files at r7, 7 of 7 files at r8.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/rapid_restart_test.go, line 58 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxhzJYr72BWZHVuPzV0:-KxhzJYr72BWZHVuPzV1:blkohgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L58)):*
> ```Go
> 	if err == nil {
> 		// Server shut down cleanly. Note that returning `err` here would create
> 		// an error interface wrapping a nil *ExitError, which is *not* nil
> ```

For this reason it's probably a better idea to rename `err` to `exitErr`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/rapid_restart_test.go, line 33 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-Kxf_HJu_rayJqwiPU8w:-KxiYe6mfPVfJWc1Pz2A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L33)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

move up
</blockquote></details>

Done.

---

*[pkg/acceptance/rapid_restart_test.go, line 58 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxhzJYr72BWZHVuPzV0:-KxiTWlp4ntLo1fMisNZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L58)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

For this reason it's probably a better idea to rename `err` to `exitErr`. 
</blockquote></details>

Done.

---

*[pkg/acceptance/rapid_restart_test.go, line 71 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfaiF83icnIhEl259O:-KxiTzhPW-e-suEeC6If:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L71)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: I think this could be cleaned up using the following structure.
```
status := s.ExitStatus()
switch status {
case -1:
    // -1 corresponds to ...
case 1:
    // 1 corresponds to ...
default:
    return ...
}
```
</blockquote></details>

Done.

---

*[pkg/acceptance/rapid_restart_test.go, line 84 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfcbQV9Z_wNjRyivxv:-KxiU-TQumsuWVA-CKd3:br90yww) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L84)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you think there's value in having one of these tests for a multi-node cluster?
</blockquote></details>

It wouldn't hurt, but it's more than I've set out to do here. Also, the acceptance tests tend to not be set up to return early, so it might require some more refactors.

---

*[pkg/acceptance/rapid_restart_test.go, line 103 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfbUhHk5xK3CepCWvg:-KxiULM3mtUgOovrDiOb:b2x3o2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L103)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Isn't this TODO already covered/discussed in `unexpectedExitCode`?
</blockquote></details>

Yup, forgot to remove this. Done.

---

*[pkg/acceptance/rapid_restart_test.go, line 122 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfcNDacObL0SixWEno:-KxiUU1IK2IGN6fZU3XE:b-fe175f) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/rapid_restart_test.go#L122)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We don't want to do this for fresh nodes?
</blockquote></details>

Sure, why not. Added.

---

*[pkg/acceptance/version_upgrade_test.go, line 69 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfcjsOzIj4aRpNuqjD:-KxiXBQoqQ4mfMWy1zTz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/version_upgrade_test.go#L69)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

super nit: verify -> verifying
</blockquote></details>

Done.

---

*[pkg/acceptance/localcluster/cluster.go, line 706 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfdQEG60cuzOMsawoB:-KxiXS2J5LMsbCizeWrp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/localcluster/cluster.go#L706)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: this should probably have a more general name, like `readFileIfExists`
</blockquote></details>

Done.

---

*[pkg/cli/start.go, line 656 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfZ4bNdoIx8tjszSny:-KxiXVGs46tbF8ZtgRTi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/272bc0b4fe655531e8572848e695313d23028a37/pkg/cli/start.go#L656)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`noDrain` doesn't convey very much meaning when only reading the bottom of this function. Consider `stopWithoutDrain`.
</blockquote></details>

Done.

---

*[pkg/cli/start.go, line 698 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfZ_ewEomxniI0vrGF:-KxiXjoej-pZSNvJjraB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/272bc0b4fe655531e8572848e695313d23028a37/pkg/cli/start.go#L698)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a comment that checking `drainingIsSafe` here can race with it being set but that it's ok to not drain even if the `Server` has started (it's just not desirable).
</blockquote></details>

Done.

---

*[pkg/cli/start.go, line 754 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfZJFG3SEp6IhGzJQ3:-KxiXoH3Woz49kbnBR62:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/272bc0b4fe655531e8572848e695313d23028a37/pkg/cli/start.go#L754)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> server drained ...

did it? Your call if it's worth printing a different message here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 7 files at r12, 1 of 1 files at r13.
Review status: 6 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/version_upgrade_test.go, line 69 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfcjsOzIj4aRpNuqjD:-KxiagA7BMGnI9Y0lEeL:btl741m) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/version_upgrade_test.go#L69)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Probably don't need a separate commit for this :stuck_out_tongue: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-:-Kxiaoc6sDkKUgHqDIOa:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 6 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/version_upgrade_test.go, line 69 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591#-KxfcjsOzIj4aRpNuqjD:-KxicbWz7GOPKxoeESXQ:b-2qavvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d8c91100d9fce21071571cd1e9d2777d0108a38/pkg/acceptance/version_upgrade_test.go#L69)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Probably don't need a separate commit for this :stuck_out_tongue: 
</blockquote></details>

You might have a point.. let me see what I can do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19591)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTRs!
